### ISSUE TITLE
Race condition in streaming producers means producer may never be unpaused

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10529]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#811|
|Type|defect|
|Created|2004-12-07 22:25:29Z|

<details><summary>Searchable metadata</summary>

```
trac-id__811 811
type__defect defect
reporter__itamarst itamarst
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1102458329000000 1102458329000000
changetime__1166859994000000 1166859994000000
version__ 
owner__ 
cc__itamarst cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
&lt;itamar> if you get a resumeProducing, write a lot of data, which then causes
the producer to be paused again
&lt;itamar> the producer will never get unpaused
&lt;itamar> cause self.producerPaused = 0 is called after resumeProducing()
&lt;itamar> resumeProducing() calls producer, producer writes, pauseProducing() and
producerPaused = 1 are set
&lt;itamar> and then the producerPaused = 0
&lt;itamar> so now transport thinks producer is not paused, even though it is
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
any progress? any test cases? pls thx
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
No tests, so leaving as bug, but James fixed it at least.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

This will be resolved by [#2286](https://github.com/twisted/twisted/issues/2286), which adds a test for the bug which was fixed here.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph set status to `closed`|
|-|-|

(In [19034]) Fix FileDescriptor to not call pauseProducing on non-streaming producers.

This fixes a bug in FileDescriptor's interaction with producers where it would
call pauseProducing() on streaming producers, which are not supposed to be
required to implement such a method.

It also adds a test for the bug in [#10529](https://github.com/twisted/twisted/issues/10529), and verifies that streaming producers
will be correctly unpaused in all cases.

Author: zseil, exarkun

Reviewer: glyph

Fixes [#2286](https://github.com/twisted/twisted/issues/2286)
Fixes [#10529](https://github.com/twisted/twisted/issues/10529)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
